lex.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) HashiCorp, Inc.
+
+package mql
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+type lexStateFunc func(*lexer) (lexStateFunc, error)
+
+type lexer struct {
+	source  *bufio.Reader
+	current stack[rune]
+	tokens  chan token
+	state   lexStateFunc
+}
+
+func newLexer(s string) *lexer {
+	l := &lexer{
+		source: bufio.NewReader(strings.NewReader(s)),
+		state:  lexStartState,
+		tokens: make(chan token, 1), // define a ring buffer for emitted tokens
+	}
+	return l
+}
+
+// nextToken is the external api for the lexer and it simply returns the next
+// token or an error. If EOF is encountered while scanning, nextToken will keep
+// returning an eofToken no matter how many times you call nextToken.
+func (l *lexer) nextToken() (token, error) {
+	for {
+		select {
+		case tk := <-l.tokens: // return a token if one has been emitted
+			return tk, nil
+		default: // otherwise, keep scanning via the next state
+			var err error
+			if l.state, err = l.state(l); err != nil {
+				return token{}, err
+			}
+
+		}
+	}
+}
+
+// lexStartState  is the start state.  It doesn't emit tokens, but rather
+// transitions to other states.  Other states typically transition back to
+// lexStartState after they emit a token.
+func lexStartState(l *lexer) (lexStateFunc, error) {
+	panicIfNil(l, "startState", "lexer")
+	r := l.read()
+	switch {
+	// wait, if it's eof we're done
+	case r == eof:
+		l.emit(eofToken, "")
+		return lexEofState, nil
+
+	// start with finding all tokens that can have a trailing "="
+	case r == '>':
+		return lexGreaterState, nil
+	case r == '<':
+		return lexLesserState, nil
+
+		// now, we can just look at the next rune...
+	case r == '%':
+		return lexContainsState, nil
+	case r == '=':
+		return lexEqualState, nil
+	case r == '!':
+		return lexNotEqualState, nil
+	case r == ')':
+		return lexRightParenState, nil
+	case r == '(':
+		return lexLeftParenState, nil
+	case isSpace(r):
+		return lexWhitespaceState, nil
+	default:
+		l.unread()
+		return lexStringState, nil
+	}
+}
+
+// lexStringState scans for strings and can emit the following tokens:
+// orToken, andToken, containsToken, stringToken
+func lexStringState(l *lexer) (lexStateFunc, error) {
+	panicIfNil(l, "lexStringState", "lexer")
+	defer l.current.clear()
+
+	// we'll push the runes we read into this buffer and when appropriate will
+	// emit tokens using the buffer's data.
+	var buf bytes.Buffer
+
+	// before we start looping, let's found out if we're scanning a quoted string
+	r := l.read()
+	var quotedString bool
+	switch r {
+	case '"':
+		quotedString = true
+	default:
+		l.unread()
+	}
+
+WRITE_TO_BUF:
```

Comment:
```suggestion
WriteToBuf:
```
I think the [convention](https://go.dev/doc/effective_go#mixed-caps) is to use mixed caps for [label names](https://stackoverflow.com/questions/64039172/whats-the-standard-practice-for-naming-labels-in-go).